admin/worker/upgrade_latest_version_projects.go

```diff
@@ -0,0 +1,93 @@
+package worker
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-version"
+	"github.com/rilldata/rill/admin"
+	"github.com/rilldata/rill/admin/database"
+	"github.com/rilldata/rill/runtime/pkg/observability"
+	"go.uber.org/zap"
+)
+
+func (w *Worker) upgradeLatestVersionProjects(ctx context.Context) error {
+	// Resolve latest version
+	latestVersion := "latest"
+	if w.admin.VersionNumber != "" {
+		latestVersion = w.admin.VersionNumber
+	}
+
+	// Verify latest version is a valid SemVer or 'latest'
+	if latestVersion != "latest" {
+		_, err := version.NewVersion(latestVersion)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Iterate over batches of projects with 'latest' version
+	limit := 100
+	afterName := ""
+	stop := false
+	for !stop {
+		// Get batch and update iterator variables
+		projs, err := w.admin.DB.FindProjectsByVersion(ctx, "latest", afterName, limit)
+		if err != nil {
+			return err
+		}
+		if len(projs) < limit {
+			stop = true
+		}
+		if len(projs) != 0 {
+			afterName = projs[len(projs)-1].Name
+		}
+
+		// Process batch
+		for _, proj := range projs {
+			err := w.upgradeAllDeploymentsForProject(ctx, proj, latestVersion)
+			if err != nil {
+				// We log the error, but continues to the next deployment
+				w.logger.Error("upgrade latest version projects: failed to upgrade project deployments", zap.String("project", proj.ID), zap.String("version", latestVersion), observability.ZapCtx(ctx))
```

Comment:
Use `project_id` instead of `project` for the key – for consistency with other logs, so searching is easier :)

Reply:
Updated